framework/src/main/java/org/checkerframework/framework/util/QualifierKindHierarchy.java

```diff
@@ -0,0 +1,839 @@
+
+    ///////////////////////////////////////////////////////////////////////////
+    /// <editor-fold desc="Initialize hierarchy">
+    /// Methods that initialize the hierarchy; only called at object initialization.
+    ///
+
+    /**
+     * Creates a {@link QualifierKindHierarchy}. Also, creates and initializes all its qualifier
+     * kinds.
+     *
+     * @param qualifierClasses all the classes of qualifiers supported by this hierarchy
+     */
+    public QualifierKindHierarchy(Collection<Class<? extends Annotation>> qualifierClasses) {
+        this.nameToQualifierKind = createQualifierKinds(qualifierClasses);
+        Map<QualifierKind, Set<QualifierKind>> directSuperMap = createDirectSuperMap();
+        specifyBottom(directSuperMap, null);
+        this.tops = createTopsSet(directSuperMap);
+        this.bottoms = createBottomsSet(directSuperMap);
+        this.topToPoly = initializePolymorphicQualifiers();
+        initializeQualifierKindFields(directSuperMap);
+        this.lubs = createLubsMap();
+        this.glbs = createGlbsMap();
+
+        verifyHierarchy(directSuperMap);
+    }
+
+    /**
+     * Verifies that the {@link QualifierKindHierarchy} is a valid hierarchy.
+     *
+     * @param directSuperMap mapping from qualifier to its direct supertypes; used to verify that a
+     *     polymorphic annotation does not have a {@link SubtypeOf} meta-annotation
+     * @throws UserError if the hierarchy isn't valid
+     */
+    protected void verifyHierarchy(Map<QualifierKind, Set<QualifierKind>> directSuperMap) {
+        for (QualifierKind qualifierKind : nameToQualifierKind.values()) {
+            boolean isPoly = qualifierKind.isPoly;
+            boolean hasSubtypeOfAnno = directSuperMap.containsKey(qualifierKind);
+            if (isPoly && hasSubtypeOfAnno) {
+                // This is currently not supported. At some point we might add
+                // polymorphic qualifiers with upper and lower bounds.
+                throw new UserError(
+                        "AnnotatedTypeFactory: "
+                                + qualifierKind
+                                + " is polymorphic and specifies super qualifiers.%n"
+                                + "Remove the @PolymorphicQualifier or @SubtypeOf annotation from it.");
+            } else if (!isPoly && !hasSubtypeOfAnno) {
+                throw new UserError(
+                        "AnnotatedTypeFactory: %s does not specify its super qualifiers.%n"
+                                + "Add an @SubtypeOf or @PolymorphicQualifier annotation to it,%n"
+                                + "or if it is an alias, exclude it from `createSupportedTypeQualifiers()`.%n",
+                        qualifierKind);
+            } else if (isPoly) {
+                if (qualifierKind.top == null) {
+                    throw new UserError(
+                            "PolymorphicQualifier, %s,  has to specify type hierarchy, if more than one exist; top types: [%s] ",
```

Comment:
Be specific about the needed fix, similarly to the errors in the previous two cases.